### PR TITLE
✨ Vendor secrets and session oenvs (part 3) 🗃️

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

This PR completes the implementation of *vendor secrets* and *session* *oenvs* (osparc-environents) but it **still does not active** in the dynamic/computational services. This will be done in two separated PRs for clarity.

A small review of what was implemented:
- A ``Template`` substitution mechanism was introduced in https://github.com/ITISFoundation/osparc-simcore/pull/3803
- More session information was need for the dynamic services. This was passed to ``ScheduleData``, specifically ``current_product``, in  https://github.com/ITISFoundation/osparc-simcore/pull/3843
- This PR implements on top of those two. It introduces  **osparc-environments** (*oenvs* in short) are identifiers-value maps that are substituted on the service specs (e.g. docker-compose).
   - **vendor secrets**: information set by a vendor on the platform. e.g. a vendor service license
   - **session oenvs**: some session information as "current user email" or the "current product name"
   - **lifespan oenvs**: produced  before a service is started and cleaned up after it finishes (e.g. API tokens ). 
- NOTE: that neither computational nor dynamic services still substitutes oenvs (will happen in subsequest PRs)

### Changes highlights
 - 🗃️ new table `services_vendor_environments` for vendor-services
    - see ``utils_services_environments`` as well 
    - NOTE that timestamps are timezone-aware (see https://github.com/ITISFoundation/osparc-simcore/issues/3131)
image

 - new ``oenvs_substitutions`` module in ``director_v2``. 
   -  see oevns registration mechanism in ``oenvs_substitutions.setup``

### Follow up
- Sync with @GitHK  to activate this functionality in dynamic services (discussion ongoing)
- Sync with @sanderegg to activate this functionality in computational services 
- @colinRawlings @mguidon please check `packages/models-library/tests/test_service_settings_labels.py:service_labels` fixture to see if this covers your user case.

## Related issue/s

-  #3565 : implements vendor and session environments

### CODE REVIEW

**Comment by colinRawlings** [comment]
# Context

This would be very helpful for crash reporting in `s4l:web`.  In this use case we have a set of services:
- `s4l-core` (the backend)
- `rt-web` (the frontend)
- `sym-server` (the symbol server)

The `sym-server` holds the "sensitive" (in the sense that it indirectly exposes the source code) symbol database.  When a crash occurs `s4l-core`'s `monitor` application `POST` the binary `.dmp` file for analysis using the symbol database.  Following analysis the callstack and metadata should be placed in an AWS SQS queue for later ingestion by the gitlab issue tracker.  The `sym-server` obtains the details for the AWS queue from (3) environment variables.

# As a (developer-)user I would like:
- A way to inject the "secret" environment variables for the queue into the service (this seems consistent with the `vendor oenvs`).  This would avoid me having to add them in the code and allow me to replace all of them everywhere if a set of credentials was comprimised without changing the code or deployed version.

- A way to learn to augment the meta-data for each crash with the user info, platform info, ... (this seems consistent with the `session oenvs`)

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 1]
```diff
@@ -149,6 +161,10 @@ class Config(_BaseConfig):
 
 SimcoreServiceSettingsLabel = ListModel[SimcoreServiceSettingLabelEntry]
 
+#
+# "simcore.service.paths-mapping" label
+#
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:166]: why don't you put these as docstring instead of a flying comment? it is already separating itself form the class...

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 2]
```diff
@@ -240,14 +256,17 @@ class Config(_BaseConfig):
...
+ComposeSpecLabelDict: TypeAlias = dict[str, Any]
 
 
 class RestartPolicy(str, Enum):
     NO_RESTART = "no-restart"
     ON_INPUTS_DOWNLOADED = "on-inputs-downloaded"
 
 
+#
+# "simcore.service.containers-allowed-outgoing-permit-list" label
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:268]: same

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 3]
```diff
@@ -56,6 +65,9 @@ class Config(_BaseConfig):
         }
 
 
+#
+# "simcore.service.settings" label
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:69]: same

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 4]
```diff
@@ -298,6 +317,11 @@ def iter_tcp_ports(self) -> Iterator[PortInt]:
                 yield port
 
 
+#
+#  "simcore.service.*" labels
```
Comment (sanderegg) [inline packages/models-library/src/models_library/service_settings_labels.py:321]: same

File: packages/models-library/tests/test_service_settings_labels.py [hunk 5]
```diff
@@ -387,3 +392,189 @@ def test_not_allowed_in_both_permit_list_and_outgoing_internet():
         f"Not allowed common_containers={{'{container_name}'}} detected"
         in f"{exec_info.value}"
     )
+
+
+@pytest.fixture
+def vendor_environments() -> dict[str, Any]:
+    return {
+        # NO list or dict!
```
Comment (sanderegg) [inline packages/models-library/tests/test_service_settings_labels.py:400]: not sure I get this comment. what does that mean? especially since there is a list right after...

File: packages/models-library/tests/test_service_settings_labels.py [hunk 6]
```diff
@@ -387,3 +392,189 @@ def test_not_allowed_in_both_permit_list_and_outgoing_internet():
...
+        "OSPARC_ENVIRONMENT_VENDOR_SECRET_TCP_PORTS": [
+            1,
+            2,
+            3,
+            4,
+        ],
+        # NOT ALLOWED since x="$QUOTED_IDENTIFIER" -> x="[1, 2, 3, 4]"
+        # enforces ALWAYS a string value before pydantic parsing and pydantic
+        # can only cast primitive types upon parse
+        # Using Json[T] would be an alternative but there are some edge cases
```
Comment (sanderegg) [inline packages/models-library/tests/test_service_settings_labels.py:413]: not sure that comment is very useful to the env pagan that I am...

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+revision = "ef817566287e"
+down_revision = "0c084cb1091c"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.create_table(
+        "services_vendor_secrets",
+        sa.Column("service_key", sa.String(), nullable=False),
```
Comment (sanderegg) [inline packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py:22]: no relation to services_meta_data? why?

File: services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py [hunk 8]
```diff
@@ -352,10 +352,11 @@ def make(cls, node_uuid: UUID) -> "DynamicSidecarNamesHelper":
         )
 
 
-class SchedulerData(CommonServiceDetails, DynamicSidecarServiceLabels):
+class SchedulerData(CommonServiceDetails):
     # TODO: ANE this object is just the context of the dy-sidecar. Should
     # be called like so and subcontexts for different handlers should
     # also be added. It will make keeping track of env vars more easily
+    DynamicSidecarServiceLabels
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py:359]: hmm what is this?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py [hunk 9]
```diff
@@ -138,10 +139,10 @@ def _update_resource_limits_and_reservations(
 
         if docker_compose_major_version >= 3:
             # compos spec version 3 and beyond
-            deploy = spec.get("deploy", {})
-            resources = deploy.get("resources", {})
-            limits = resources.get("limits", {})
-            reservations = resources.get("reservations", {})
+            deploy: dict[str, Any] = spec.get("deploy", {})
+            resources_v3: dict[str, Any] = deploy.get("resources", {})
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:143]: _v3???

**Review by sanderegg — COMMENTED** [review]
Seems good. please check the database relations. why you make service_key a free field?

File: packages/models-library/tests/test_service_settings_labels.py (hunk 5 above)
Comment (pcrespov) [reply packages/models-library/tests/test_service_settings_labels.py:400]: Sorry! Forgot to update this comment!

File: services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py (hunk 8 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/models/schemas/dynamic_services/scheduler.py:359]: oohh sorry for that... was a bad commit! 
I wrote that there while discussing with ANE ... the linter caught it but I did not notice it!

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py (hunk 9 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:143]: this is to highlight that from dcoker-compose > 3, `limits` and `reservations` etc are grouped under `resources`. `resources` is a very generic name and i thought this might highlight that. In addition, this entry is also mapped into `deploy.resources` of the `specs`, so it has "double use"

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py (hunk 7 above)
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py:22]: In short: this is not possible with the current design. we would have to normalize the services_metadata table.
More details:

- `services_vendor_secrets` has only `key` column but cannot use a foreign key to `services_metadata` because the latter does not have a unique `key`. Note that the later is unique on (key, version) pair. Therefore there is not a unique correspondence. 
- A way to achieve this is normalizing the current design, i.e. create separately a table for service keys, then another for service versions and then use service_metadata as a mapping table. 
```mermaid
classDiagram
    class services {
        key
    }

    class versions {
        tag
    }

    class services_metadata {
        service_key
        service_version
        description
    }

    services_metadata "1" -- "1" services : key
    services_metadata "1" -- "1" versions : tag
```
NOTE: perhaps the versions table is not necessary .

Then we could use a foreign key from ``services_vendor_secrets` to the new `services` table
```mermaid
classDiagram
    class services {
        key
    }

    class versions {
        tag
    }

    class services_metadata {
        service_key
        service_version
        description
    }
    
    class services_vendor_secrets {
        service_key
        vendor_secrets
    }

    services_metadata "1" -- "1" services : key
    services_metadata "1" -- "1" versions : tag
    services_vendor_secrets "1" -- "1" services : key

```

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py (hunk 9 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:143]: but then probably you should rename more than half of the code in there.. not sure that make sense. I think the deploy part is also very much changed from v2 to v3. do we even support v2? not sure. maybe we should better define this and then if we only support v3, which I think is probably the case, remove it.

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py (hunk 7 above)
Comment (sanderegg) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py:22]: well that is interesting, I did not know about that naming and it makes sense. So if I get your problem is that you cannot create a relation to the key because these are not unique in the services_metadata. I think we do not need the versions table until we need it. Then comes the second question, why you do not have the versions in your vendor_secrets table? what if some ENV is only necessary from some version onward? what if I want to change it but only from version 2.x.y? should you not have the version in there?

Also should we not finally put the semver plugin in postgres?

Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py:22]: i have mixed fillings about versions. IMO it requires too much maintenance and there are very few cases in which some properties changes from version to version within the table. I would rather use version to define "boundaries". Something like: from the version on ... these are the properties. Then the table will include "ranges":


| service | version | color |
| - | - | - |
| A | 5.0 | red |
| A | 1.1 | blue |
| A | 1.0 | green |


means that service `A` only was green between 1.0 and 1.1, then blue and in 5.0 (latest) it turned red. 
This obviously assumes that:
- service A has published all versions ranges in between (which might not be true)
-

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py (hunk 9 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:143]: yes, if you see the entire code, there is a part that supports v2

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py (hunk 7 above)
Comment (sanderegg) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py:22]: well. because of this you cannot use the existing relation to services_meta_data.
Also I can see things like:
s4l-lite:2.1.343 needs ENV XXX_BLAHBLAH
from s4l-lite:3.1.2 needs not XXX_BLAHBLAH but also YYY_BLIHBLAHBLOUH
I agree that in this case, it is probaly harmless to pass the same to 2.1.343.
But what if the use-case is like so:
- SOME_HOST=http://some_url
and from version 3.1.2 it needs to be another protocol such as tcp://blahblahblah ??
In the second one then you are doomed.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py (hunk 9 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:143]: yes but my point is that the whole deploy part is probably v3 only https://nickjanetakis.com/blog/docker-tip-51-which-docker-compose-api-version-should-you-use#:~:text=Some%20of%20the%20major%20differences,ll%20need%20to%20use%20v3%2B.
do we really support v2?

File: packages/models-library/tests/test_service_settings_labels.py [hunk 10]
```diff
@@ -387,3 +392,183 @@ def test_not_allowed_in_both_permit_list_and_outgoing_internet():
         f"Not allowed common_containers={{'{container_name}'}} detected"
         in f"{exec_info.value}"
     )
+
+
+@pytest.fixture
```
Comment (GitHK) [inline packages/models-library/tests/test_service_settings_labels.py:397]: why not keep all these fixtures and tests for next PR since they are skipped now. It will be simpler to review

File: packages/models-library/tests/test_utils_string_substitution.py [hunk 11]
```diff
@@ -162,15 +164,87 @@ def test_substitution_with_new_and_legacy_identifiers():
...
+    y=3
+    """
+    )
+
+    # NOTE does not cast if it is in a container
+    assert template.substitute({"VALUE1": ["3", "4"], "VALUE2": [3, 4]}) == dedent(
+        """
+    x=['3', '4']
+    y=[3, 4]
+    """
```
Comment (GitHK) [inline packages/models-library/tests/test_utils_string_substitution.py:203]: From what I get here, python values are serialised to string. Did I get it right?

File: packages/postgres-database/src/simcore_postgres_database/utils_services_environments.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+from ._protocols import DBConnection
+from .models.services_environments import VENDOR_SECRET_PREFIX, services_vendor_secrets
+
+# This constraint is to avoid deserialization issues after substitution!
+VendorSecret: TypeAlias = bool | int | float | str
+
+
+async def get_vendor_secrets(
+    conn: DBConnection,
+    vendor_service_key: str,  # NOTE: ServiceKey is defined in model_library
```
Comment (GitHK) [inline packages/postgres-database/src/simcore_postgres_database/utils_services_environments.py:14]: Does it complain if you try to do something like this?
```suggestion
    vendor_service_key: "ServiceKey",  # NOTE: ServiceKey is defined in model_library
```
```suggestion
    vendor_service_key: str,  # NOTE: ServiceKey is defined in model_library
```

File: packages/postgres-database/src/simcore_postgres_database/utils_services_environments.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    if normalize_names:
+        for key in list(secrets.keys()):
+            if not key.startswith(VENDOR_SECRET_PREFIX):
+                secrets[VENDOR_SECRET_PREFIX + key.upper()] = secrets.pop(key)
+
+        assert all(key.startswith(VENDOR_SECRET_PREFIX) for key in secrets)  # nosec
+
+    assert all(  # nosec
+        isinstance(value, (bool, int, str, float)) for value in secrets.values()
+    )
```
Comment (GitHK) [inline packages/postgres-database/src/simcore_postgres_database/utils_services_environments.py:37]: Why not make this a check and raise a proper error at this point. I would guess this should be good to have at runtime as well.

File: packages/postgres-database/tests/test_utils_services_environments.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+                for key, value in vendor_secrets.items()
+            },
+        )
+    )
+
+    return vendor_secrets
+
+
+def test_vendor_secret_prefix_has_underscore():
+    assert VENDOR_SECRET_PREFIX.endswith("_")  # should allow
```
Comment (GitHK) [inline packages/postgres-database/tests/test_utils_services_environments.py:77]:
```suggestion
def test_vendor_secret_prefix_must_end_with_underscore():
    assert VENDOR_SECRET_PREFIX.endswith("_")
```
```suggestion
def test_vendor_secret_prefix_has_underscore():
    assert VENDOR_SECRET_PREFIX.endswith("_")  # should allow
```

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/services_environments.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        """Fetches vendor secrets for a service using normalized names"""
+        async with self.db_engine.acquire() as conn:
+            vendor_secrets: dict[str, VendorSecret] = await get_vendor_secrets(
+                conn, vendor_service_key=service_key, normalize_names=True
+            )
+
+            # NOTE: normalize_names = True
+            assert all(  # nosec
+                self.is_vendor_secret_identifier(key) for key in vendor_secrets
+            )  # nosec
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/db/repositories/services_environments.py:31]: I would also transform this in a check rather than have an assert

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py (hunk 9 above)
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:143]: yes V2 is also supported, in the else block below.
I would not rename the variable here. The `if else` statement encapsulates this. At most create 2 functions to wrap this code for v2 and v3

File: services/director-v2/tests/unit/test_utils_session_oenvs.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    async def request_user_email(app: FastAPI, user_id: UserID) -> SubstitutionValue:
+        print(app, user_id)
+        await asyncio.sleep(1)
+        return faker.email()
+
+    # Some context given ----------------------------------------------------------
+    # TODO: test pre errors handling
+    # TODO: test errors handling
+    # TODO: test validation errors handling
+    # TODO: test timeout error handling
```
Comment (GitHK) [inline services/director-v2/tests/unit/test_utils_session_oenvs.py:85]: Can these be addressed before they are merged? Otherwise keep them not committed for the next PR.

**Review by GitHK — APPROVED** [review]
In general just some minor things. And stuff I would like to see moved to the next PR.

File: packages/models-library/tests/test_service_settings_labels.py (hunk 10 above)
Comment (pcrespov) [reply packages/models-library/tests/test_service_settings_labels.py:397]: Some reasons are:
- TDD: i prefer to have tests ready before the features are implemented
- They are also illustrative of user-cases for follow-up PR
- I do not forget details (in case I cannot do the follow-up right away)

File: packages/models-library/tests/test_utils_string_substitution.py (hunk 11 above)
Comment (pcrespov) [reply packages/models-library/tests/test_utils_string_substitution.py:203]: its not strictly about serialization but about substitution

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/services_environments.py (hunk 15 above)
Comment (pcrespov) [reply services/director-v2/src/simcore_service_director_v2/modules/db/repositories/services_environments.py:31]: This is guaranteed by the utils function

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py (hunk 9 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py:143]: let it be.

**Review by sanderegg — APPROVED** [review]
discussed together.
- add service_base_version
- make it work like services_access_rights and services_specifications table

File: packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py (hunk 7 above)
Comment (pcrespov) [reply packages/postgres-database/src/simcore_postgres_database/migration/versions/ef817566287e_new_services_vendor_secrets_table.py]: pair reviewed this changes with @sanderegg @matusdrobuliak66
